skstore/src/Utils.sk

```diff
@@ -0,0 +1,230 @@
+
+fun makeInput<K: Key, V: File>(
+  convKey: Key ~> K,
+  convValue: File ~> V,
+  context: mutable Context,
+  name: String,
+  content: Array<(K, V)> = Array[],
+): String {
+  dirName = subDirName(context, name);
+  _ = context.mkdir(convKey, convValue, dirName, content);
+  dirName.toString()
+}
+
+fun makeLazy<K: Key, V: File>(
+  convKey: Key ~> K,
+  convValue: File ~> V,
+  context: mutable Context,
+  name: String,
+  f: (mutable Context, LHandle<K, V>, K) ~> Array<V>,
+  collect: Bool = true,
+): String {
+  dirName = subDirName(context, name);
+  _ = LHandle::create(convKey, convValue, context, dirName, f, collect);
+  dirName.toString()
+}
+
+class Check<V: File>(valid: Bool, optPrevious: ?V)
+class Loading<V: File>(internal: V, result: V)
+
+fun makeAsyncLazy<K: Key, V: File, C: frozen>(
+  convKey: Key ~> K,
+  convValue: File ~> V,
+  context: mutable Context,
+  name: String,
+  params: (mutable Context, K) ~> C,
+  call: (String, String, K, C) ~> void,
+  check: (V, C) ~> Check<V>,
+  loading: (String, C, ?V) ~> Loading<V>,
+): String {
+  dirName = subDirName(context, name);
+  resultsDirs = DirName::create(dirName.toString() + "sk_prv/results/");
+  results = context.mkdir(convKey, convValue, resultsDirs, Array[]);
+  _ = LHandle::create(convKey, convValue, context, dirName, (ctx, self, key) ~> {
+    ps = params(ctx, key);
+    // FastOption error: Signature must be specified
+    perform: (?V) -> Array<V> = optPrevious -> {
+      ksuid = Ksuid::create().toString();
+      call(self.getDirName().toString(), ksuid, key, ps);
+      loadingValue = loading(ksuid, ps, optPrevious);
+      ctx.postpone(
+        ToWrite(resultsDirs, key, _ctx ~> Array[loadingValue.internal]),
+      );
+      Array[loadingValue.result]
+    };
+    results.maybeGet(ctx, key).map(r -> check(r, ps)) match {
+    | Some(Check(true, Some(v))) -> Array[v]
+    | Some(Check(_, optPrevious)) -> perform(optPrevious)
+    | _ -> perform(None())
+    }
+  });
+  dirName.toString()
+}
+
+fun accReducer<V1: File, V2: File>(
+  accumulator: Accumulator<V1, V2>,
+): EReducer<V1, V2> {
+  EReducer{
+    type => accumulator.getType(),
+    canReset => true,
+    init => iter ~> {
+      acc = accumulator.default;
+      for (x in iter) {
+        !acc = accumulator.accumulate(acc, x)
+      };
+      Array[acc]
+    },
+    update => (state, old, new) ~> {
+      acc = state[0];
+      for (x in old) {
+        accumulator.dismiss(acc, x) match {
+        | Some(v) -> !acc = v
+        | _ -> return None()
+        }
+      };
+      for (x in new) {
+        !acc = accumulator.accumulate(acc, x)
+      };
+      Some(Array[acc])
+    },
+  }
+}
+
+fun subDirName(context: mutable Context, name: String): DirName {
+  path = context.currentArrow() match {
+  | Some(arrow) ->
+    b64 = base64(arrow.key);
+    `${arrow.childName}${b64}/${name}/`
+  | None() -> `/sk/${name}/`
+  };
+  DirName::create(path)
+}
+
+fun parentName(context: mutable Context): DirName {
```

Comment:
What is this used for? It also seems to fix a convention of `/sk/` as the "global" parent (along with the line 207 case above) -- what's the story there?

Reply:
Useless function from older version. Indeed, as we mixing SKDB tables directories and SKStore directories and as the directory name are fully generated, I choose to have a convention of '/sk/' as the "global" parent for the pure SKStore side, just a habit for mixing skstore computation graphs.